### ISSUE TITLE
Unble to get locations in the android app

### ISSUE THREAD

**Issue by Mayur2502** [aelassas/bookcars#43]
i have run the app but on the index screen i am unable to get locations it doesnt show any suggestions and i have sucessfully run api and connected the database also so help me with the above concen

**Comment by aelassas** [comment]
Did you update BC_API_HOST in mobile/.env?

You should put your IP address if you are on a dev environment.

Make sure that the API port is not blocked by the firewall.

You should update BC_CDN_USERS and BC_CDN_CARS as well.

Then run the mobile app with the following command:
```
yarn start:clean
```

This is described [here](https://github.com/aelassas/bookcars/wiki/Run-Mobile-App).

**Comment by Mayur2502** [comment]
Yes I have updated api host  but when i start node server it is start on
4002 port and when I check the 4002 port in browser it shows cannot get/
can u elaborate what Is the issue
I am new to the devlopement and u help means a lot to me Thank you so much

**Comment by aelassas** [comment]
First, on your local computer, try the following query: http://localhost:4002/api/locations/1/30/en

Then, on your mobile device, try the following query: http://IP:4002/api/locations/1/30/en

Replace **IP** with your local computer IP address. If you're on windows you can get your IP with `ipconfig` command.

**Comment by Mayur2502** [comment]
I have run both the LINKS(query) in the browser and after running both both
the commands it shows like this
[image: image.png]

**Comment by aelassas** [comment]
I can't see your image. You should add it from GitHub.

**Comment by Mayur2502** [comment]
https://github.com/Mayur2502/pw_skill-assignment/blob/40ee0e3aebf64ad9e1d25ecc815cf49b6ef1c6dd/image.png

**Comment by Mayur2502** [comment]
![Screenshot 2023-08-24 001545](https://github.com/aelassas/bookcars/assets/124443601/697df27a-7405-407a-8f4d-2adc0932ce7b)

**Comment by aelassas** [comment]
This means that you have no locations in your database. You can create locations from the backend or restore the [demo database](https://github.com/aelassas/bookcars/wiki/Demo-Database).

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/9c049e7e-ff73-474e-966a-683d572c3570)

okk help me with the how to restore it i am unable understand the ur steps for restoring the database
can u explain in detail

**Comment by aelassas** [comment]
First, follow the [prerequisites](https://github.com/aelassas/bookcars/wiki/Run-from-Source#prerequisites) to set up the admin user with a **password** and secure MongoDB.

Then, use the same **password** in `mongorestore` command (replace `$PASSWORD` with your MongoDB password).

**Comment by aelassas** [comment]
@Mayur2502 Are you still facing some issues?

**Comment by Mayur2502** [comment]
yes but i am trying

**Comment by Mayur2502** [comment]
where should i unzip the demo db file

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/491fd2b6-1a3d-4902-a81b-727029c642cc)

**Comment by aelassas** [comment]
@Mayur2502 `mongorestore` is a command line tool so you can't run it from mongosh. Download [MongoDB database tools](https://www.mongodb.com/try/download/database-tools).

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/a7ab194c-7dd3-4af3-b5df-d0c9ff203d50)

**Comment by aelassas** [comment]
Did you restore the demo db?

What's the result of http://localhost:4002/api/locations/1/30/en now?

**Comment by Mayur2502** [comment]
Done Bro!!! Got it Thank you so MUCH for your help Means a lot!!!!  Lastly I want to try to book a ride  and i have tried  it shows like this

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/8e89626f-38e8-40e2-b40d-07fda505eda8)

**Comment by aelassas** [comment]
Did you set up smtp settings in api/.env?

**Comment by Mayur2502** [comment]
NO i dont know anything about itt

**Comment by aelassas** [comment]
You need to create an account on mailjet or brevo or sendgrid, and set the following settings in api/.env:
```
BC_SMTP_HOST = in-v3.mailjet.com
BC_SMTP_PORT = 587
BC_SMTP_USER = USER
BC_SMTP_PASS = PASSWORD
BC_SMTP_FROM = youremail@gmail.com
```
If mailjet isn't working try brevo.

To test smtp try to sign up from the frontend or the mobile app.

If smtp is working you should drop your bookcars db and restore the demo db again to start from a clean db without any issues.

To drop the db, open `mongosh` with the following command:
```
mongosh --host 127.0.0.1 --port 27017 -u admin -p $PASSWORD --authenticationDatabase admin
```

Replace **$PASSWORD** with your password.

Then run the following commands from mongosh:
```
db = db.getSiblingDB('bookcars')
db.dropDatabase()
```

Last but not least, If everything is working fine you need to set up the cdn server. On windows, install IIS and create C:\inetpub\wwwroot\cdn\bookcars folder. Unzip cdn.zip folder and place its content in C:\inetpub\wwwroot\cdn\bookcars. You also need to add the full access permissions on C:\inetpub\wwwroot\cdn\bookcars to the user who's running the api (generally the logged in user in windows). On Linux, you'll need to install NGINX. This is explained [here](https://github.com/aelassas/bookcars/wiki/Run-from-Source).

**Comment by Mayur2502** [comment]
I have a  run the app app sucessfully and i am able to access all the content and also restore the database only thing is remained is to setup smtp

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/d299e8dd-63d0-428f-96da-a94654f19ac1)
Thiis is the result of the  http://localhost:4002/api/locations/1/30/en  now?

**Comment by aelassas** [comment]
And the CDN server.  Don't forget it. Otherwise, you won't see all the images on the mobile app, the frontend and the backend.

Good luck for the smtp and the cdn. Let le know if you face any issue.

**Comment by Mayur2502** [comment]
BC_SMTP_HOST = in-v3.mailjet.com
BC_SMTP_PORT = 587
BC_SMTP_USER = USER
BC_SMTP_PASS = PASSWORD
BC_SMTP_FROM = youremail@gmail.com
I have created a account on brevo and I got host ,port from there what should I enter in SMTP USER, PASS ?The login details of my Brevo Account or something else

**Comment by aelassas** [comment]
Click on your name in the menu at the top right.

Click SMTP & API.

Under the SMTP tab, you will find your BC_SMTP_HOST, BC_SMTP_USER and SMTP password.

Copy and paste BC_SMTP_HOST , BC_SMTP_USER and the password into BC_SMTP_PASS to integrate it with Brevo.

BC_SMTP_FROM should be the same as BC_SMTP_USER.

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/2b1099c8-d785-4dae-b0cf-ac260145d1a1)

**Comment by Mayur2502** [comment]
which is smtp user and which is password or i have to paste the master password in the smtp pass section'

**Comment by aelassas** [comment]
smtp user is your login, smtp password is the master password and smtp host is the smtp server.

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/d67ef0de-8776-47cb-9484-032d87234896)
What about this??

**Comment by Mayur2502** [comment]
My brevo account getting suspended

**Comment by aelassas** [comment]
You need to send an activation request to ***@***.*** with your
account info. They will respond you.

Otherwise, you can try mailjet.com.

**Comment by Mayur2502** [comment]
My mailjet account also got suspended i didnt even use it dont know why it is happening??

**Comment by aelassas** [comment]
You can try the followings in this order:
1. [sendpulse.com](https://sendpulse.com) 
2. [mailersend.com](https://www.mailersend.com)
3. [smtp2go.com](https://www.smtp2go.com)
4. [sendgrid.com](https://sendgrid.com/)

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/1f4314b0-7cbe-4023-92cf-5004bb891450)
it is asking for domain and too much details 
and sites i am unale to signup

**Comment by aelassas** [comment]
Try smtp2go then sendgrid.

If nothing works try to find a free transactional email provider on google.

I have set up mailjet without any issue.

**Comment by aelassas** [comment]
@Mayur2502 Try [sendgrid.com](https://signup.sendgrid.com/). I have just signed up and set up smtp and everything works fine.

Once you sign up, create a sender identity. Then click on [Use Web API or SMTP](https://app.sendgrid.com/guide/integrate). Then click on SMTP Relay. Then create an API key. Then, update api/.env, click on verify integration and send an email from BookCars (sign up).

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/29d906ba-1bec-406e-b4c9-ce927af2315e)
in This which is smtp user and which is password

**Comment by aelassas** [comment]
Api key is user and secret key is password.

**Comment by Mayur2502** [comment]
![image](https://github.com/aelassas/bookcars/assets/124443601/320accb2-fa17-4494-9c55-2eb4137c28d7)

![image](https://github.com/aelassas/bookcars/assets/124443601/e5fcdc67-e756-47b1-ad73-e739f768a79d)

What is issue with this

**Comment by aelassas** [comment]
The SMTP 501 error occurs when you try to send an email to an invalid email address or invalid domain name. Additionally, you may run into this error from time to time due to anti-virus guards dropping your SMTP connection.

Where did you get this error? In sign up?

**Comment by Mayur2502** [comment]
Yaah at the sign up page But I think I have resolved all the errors but I have exhausted the email limit

**Comment by Mayur2502** [comment]
I just wanna know what is JWT Token???

**Comment by aelassas** [comment]
JWT is used to secure the API.

One common way to secure APIs is using token, some long-character-sequence, attached on HTTP request (usually in HTTP request header), where API server validates the token.

There are two types of tokens:

- stateful: where server save the token data (e.g. on Redis), and validates each request against this token
- stateless (like JWT): in the contrary, server save nothing. The token itself contains all information required for validation.

For some routes in the API (like backend routes), we only give access to logged in backend users.

JWT token is created when the user signs in as follows:
```js
const payload = { id: user._id }

let options = { expiresIn: JWT_EXPIRE_AT }
if (req.body.stayConnected) {
  options = {}
}

const token = jwt.sign(payload, JWT_SECRET, options)
```

Then the token is saved on client side for future API requests in the backend, the frontend and the mobile app.

For example, if we take this route:
```js
routes.route(routeNames.create).post(authJwt.verifyToken, carController.create)
```

JWT token validation is done through `authJwt.verifyToken` middleware.

`verifyToken` simply validates the JWT token available in `x-access-token` HTTP request header:
```js
import process from 'node:process'
import jwt from 'jsonwebtoken'

const JWT_SECRET = process.env.BC_JWT_SECRET

function verifyToken(req, res, next) {
  const token = req.headers['x-access-token']

  if (!token) {
    return res.status(403).send({ message: 'No token provided!' })
  }

  jwt.verify(token, JWT_SECRET, (err, decoded) => {
    if (err) {
      console.log(err)
      return res.status(401).send({ message: 'Unauthorized!' })
    }

    req.userId = decoded.id
    next()
  })
}

export default { verifyToken }
```

If you open your browser network dev tools and see a request, you will find JWT in `x-access-token` HTTP request header. You can copy/paste the token on [jwt.io](https://jwt.io/) to see all the information included in the token.

**Comment by aelassas** [comment]
@Mayur2502 Did you set up BookCars CDN?

I updated the [documentation](https://github.com/aelassas/bookcars/wiki/Demo-Database).

**Comment by Mayur2502** [comment]
I have already done that but I saw u have updated it Is there any changes in it??

**Comment by aelassas** [comment]
Ok, nice. I added cdn instructions.
